Implement errors about undefined symbols

mati865 opened the pull request:
Fixes: #20

There are two regressions right now:
```
Error: Undefined symbols found!
  Undefined symbol does_not_exist, referenced by /home/mateusz/projects/wild/wild/tests/build/archive_activation.default-f0cd5ddfa601dc84.o

Error: Undefined symbols found!
  Undefined symbol __start_foo, referenced by /home/mateusz/projects/wild/wild/tests/build/custom_section.archive-95f9e368cdd19b2a.o
  Undefined symbol __stop_foo, referenced by /home/mateusz/projects/wild/wild/tests/build/custom_section.archive-95f9e368cdd19b2a.o
  Undefined symbol __start_bar, referenced by /home/mateusz/projects/wild/wild/tests/build/custom_section.archive-95f9e368cdd19b2a.o
  Undefined symbol __stop_bar, referenced by /home/mateusz/projects/wild/wild/tests/build/custom_section.archive-95f9e368cdd19b2a.o

test integration_test::program_name_19___archive_activation_c__ ... FAILED
test integration_test::program_name_15___custom_section_c__ ... FAILED
```

I don't know about the first one yet, but the second one looks like this function should take into account start/stop symbols from `canonicalise_undefined_symbols`. I'll look into it when I get more time to understand what I'm doing 😅

davidlattimore commented:
Thanks for working on this! I think both of the problems you're hitting will likely be fixed by changing where you're checking for undefined symbols. I'd suggest looking at putting the logic somewhere near the call to `send_symbol_request` in `process_relocation`. Putting the logic there will mean that you'll only be reporting errors for undefined symbols if the undefined symbol is referenced from a live object and in the case of `--gc-sections` (default in Wild), from reachable code.

mati865 commented:
Rebased to the latest main, reworked the check, and (hopefully) improved the test a bit.

Previously I had missed `report_error`, I think that's pretty neat, except for printing only one error, which is far from ideal for printing undefined references.
I'll address that issue in a followup.

davidlattimore commented:
Nice! Thanks for doing this!
